Userland/Libraries/LibWeb/HTML/Window.cpp

```diff
@@ -1395,4 +1400,166 @@ JS::NonnullGCPtr<CustomElementRegistry> Window::custom_elements()
+                property_names.set(name.value(), AK::HashSetExistingEntryBehavior::Keep);
+        }
+        return IterationDecision::Continue;
+    });
+
+    // FIXME: Many copies here. Would be nice to be able to return Vector<String>
+    Vector<DeprecatedString> names;
+    names.ensure_capacity(property_names.size());
+    for (auto const& name : property_names) {
+        names.append(name.to_deprecated_string());
+    }
+    return names;
+}
+
+// https://html.spec.whatwg.org/#named-access-on-the-window-object
+WebIDL::ExceptionOr<JS::Value> Window::named_item_value(DeprecatedFlyString const& name) const
+{
+    // To determine the value of a named property name in a Window object window, the user agent must return the value obtained using the following steps:
+
+    // 1. Let objects be the list of named objects of window with the name name.
+    // NOTE: There will be at least one such object, since the algorithm would otherwise not have been invoked by Web IDL.
+    auto objects = const_cast<Window&>(*this).named_objects(name);
+
+    // 2. If objects contains a navigable, then:
+    if (!objects.navigables.is_empty()) {
+        // 1. Let container be the first navigable container in window's associated Document's descendants whose content navigable is in objects.
+        JS::GCPtr<NavigableContainer> container = nullptr;
+        const_cast<Window&>(*this).associated_document().for_each_in_subtree_of_type<HTML::NavigableContainer>([&](HTML::NavigableContainer& navigable_container) {
+            if (!navigable_container.content_navigable())
+                return IterationDecision::Continue;
+            if (objects.navigables.contains_slow(JS::NonnullGCPtr { *navigable_container.content_navigable() })) {
+                container = navigable_container;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+        // 2. Return container's content navigable's active WindowProxy.
+        VERIFY(container);
+        return container->content_navigable()->active_window_proxy();
+    }
+
+    // 3. Otherwise, if objects has only one element, return that element.
+    if (objects.elements.size() == 1)
+        return objects.elements[0];
+
+    // 4. Otherwise return an HTMLCollection rooted at window's associated Document,
+    //    whose filter matches only named objects of window with the name name. (By definition, these will all be elements.)
+    return DOM::HTMLCollection::create(const_cast<Window&>(*this).associated_document(), DOM::HTMLCollection::Scope::Descendants, [&name](auto& element) -> bool {
+        if ((is<HTMLEmbedElement>(element) || is<HTMLFormElement>(element) || is<HTMLImageElement>(element) || is<HTMLObjectElement>(element))
+            && (element.attribute(AttributeNames::name) == name.view()))
+            return true;
+        return element.attribute(AttributeNames::id) == name.view();
```

Comment:
I repeat this pattern in 3 places But they're all slightly different... so I'm not sure how to deduplicate it :/